### PR TITLE
Implemented 'path:' prefix for browser paths (closes #468, closes #466)

### LINKED ISSUE #466: Check browser aliases from CLI parameters before tests compiling

### LINKED ISSUE #468: Implement specifying browser paths through 'path:' prefix
Require a user to add `path:`  prefix when specifying a browser path in the CLI.

What we have now:
`.\testcafe chrome,ie,C:\Firefox\firefox.exe test.js`

What it should be:
`.\testcafe chrome,ie,path:C:\Firefox\firefox.exe test.js`

Or, for if path contains spaces:
`.\testcafe chrome,ff,path:"C:\Program Files\Internet Explorer\ie.exe" test.js`
`.\testcafe chrome,ff,path:'C:\Program Files\Internet Explorer\ie.exe' test.js`

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @inikulin. @AlexanderMoskovkin, @VasilyStrelyaev (changed error message, test names)

### CODE REVIEW

File: src/errors/runtime/message.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ export default {
     browserDisconnected:                'The {userAgent} browser disconnected. This problem may appear when a browser hangs or is closed, or due to network issues.',
     cantRunAgainstDisconnectedBrowsers: 'The following browsers disconnected: {userAgents}. Tests will not be run.',
     cantEstablishBrowserConnection:     'Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure.',
-    cantFindBrowser:                    'Unable to find the browser. "{browser}" is not a browser alias or path to an executable file.',
+    cantFindBrowser:                    'Unable to find the browser. "{browser}" is not a browser alias or string "path:${PATH}", where ${PATH} is an enquoted path to an executable file.',
```
Comment (VasilyStrelyaev) [inline src/errors/runtime/message.js]:
```
    cantFindBrowser:                    'Unable to find the browser. "{browser}" is not a browser alias or a "path:${PATH}" string, where ${PATH} is an enquoted path to an executable file.',
```

File: test/server/cli-argument-parser-test.js [hunk 2]
```diff
@@ -52,38 +52,77 @@ describe('CLI argument parser', function () {
...
                 .then(function (results) {
                     var allAliasses = Object.keys(results[0]);
                     var parser      = results[1];
 
-                    expect(parser.browsers).eql(['ie', 'chrome'].concat(allAliasses));
+                    expect(parser.browsers).eql(['ff', 'ff'].concat(allAliasses));
                 });
         });
+
+        it('Should raise an error if contains a plain path', function () {
```
Comment (VasilyStrelyaev) [inline test/server/cli-argument-parser-test.js]:
```
        it('Should raise an error if an unprefixed path is provided', function () {
```

File: test/server/cli-argument-parser-test.js [hunk 3]
```diff
@@ -52,38 +52,77 @@ describe('CLI argument parser', function () {
...
+            return assertRaisesError(
+                '/Applications/Firefox.app,ff,',
+                'Unable to find the browser. "/Applications/Firefox.app" is not a browser alias' +
+                ' or string "path:${PATH}", where ${PATH} is an enquoted path to an executable file.'
+            );
+        });
+
+        it('Should raise an error if contains an unquoted path descriptor', function () {
```
Comment (VasilyStrelyaev) [inline test/server/cli-argument-parser-test.js]:
```
        it('Should raise an error if the path descriptor is unquoted', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
Tests failed due to unstable tests on iOS.
@testcafe-build-bot \retest

File: src/cli/argument-parser.js [hunk 4]
```diff
@@ -85,6 +85,39 @@ export default class CLIArgumentParser {
         return '\n' + wordWrap(DESCRIPTION, 2, getViewPortWidth(process.stdout));
     }
 
+    static _splitBrowserArg (browserArg) {
```
Comment (inikulin) [inline src/cli/argument-parser.js]: It's `parseBrowserArg` I guess?

File: src/cli/argument-parser.js [hunk 5]
```diff
@@ -85,6 +85,39 @@ export default class CLIArgumentParser {
...
+                browsers.push(browserArg.substring(itemStartIndex, currentPos - 1));
+                itemStartIndex = currentPos;
+            }
+            else if (quotedRegionMark && delimiter === quotedRegionMark)
+                quotedRegionMark = '';
+            else if (!quotedRegionMark)
+                quotedRegionMark = delimiter;
+
+            delimetersMatch = delimetersRegExp.exec(browserArg);
+        }
```
Comment (inikulin) [inline src/cli/argument-parser.js]: It's really complicated using RegExp. Can we use simple state machine instead?

File: src/errors/runtime/message.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ export default {
     browserDisconnected:                'The {userAgent} browser disconnected. This problem may appear when a browser hangs or is closed, or due to network issues.',
     cantRunAgainstDisconnectedBrowsers: 'The following browsers disconnected: {userAgents}. Tests will not be run.',
     cantEstablishBrowserConnection:     'Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure.',
-    cantFindBrowser:                    'Unable to find the browser. "{browser}" is not a browser alias or path to an executable file.',
+    cantFindBrowser:                    'Unable to find the browser. "{browser}" is not a browser alias or a "path:${PATH}" string, where ${PATH} is an enquoted path to an executable file.',
```
Comment (inikulin) [inline src/errors/runtime/message.js]: I don't get this error message. Can we just keep it as it was before?

File: src/runner/bootstrapper.js [hunk 7]
```diff
@@ -86,8 +105,9 @@ export default class Bootstrapper {
     async createRunnableConfiguration () {
         var reporterPlugin = this._getReporterPlugin();
 
-        var tests      = await this._getTests();
-        var browserSet = await this._getBrowserConnections();
+        var browserInfo = await this._getBrowserInfo();
```
Comment (inikulin) [inline src/runner/bootstrapper.js]: We need a comment here, why we use this order of execution.

**Comment by inikulin** [comment]
\r-

**Comment by AndreyBelym** [comment]
@VasilyStrelyaev. please, check the new explanatory comment [here](https://github.com/DevExpress/testcafe/pull/475/files#diff-94fc755d5ce4b3e8e12b02ccda41563fL88).

File: src/cli/argument-parser.js [hunk 8]
```diff
@@ -85,6 +85,51 @@ export default class CLIArgumentParser {
...
+
+            case '"':
+            case '\'':
+                if (!isInQuotedPart) {
+                    isInQuotedPart = true;
+                    quoteChar      = currentChar;
+                }
+                else if (quoteChar === currentChar) {
+                    isInQuotedPart = false;
+                    quoteChar      = '';
```
Comment (inikulin) [inline src/cli/argument-parser.js]: Seems like `isInQuotedPart` var is redundant. We can set `quoteChar` to null when we are not in quote (this is actually what happens currently) and use it as flag.

File: src/cli/argument-parser.js [hunk 9]
```diff
@@ -85,6 +85,51 @@ export default class CLIArgumentParser {
         return '\n' + wordWrap(DESCRIPTION, 2, getViewPortWidth(process.stdout));
     }
 
+    static _parseBrowserArg (browserArg) {
+        var browser  = '';
```
Comment (inikulin) [inline src/cli/argument-parser.js]: Let' rename `browser` to `currentBrowser` to avoid confusion with `browsers`

**Comment by AndreyBelym** [comment]
fpr

File: src/runner/bootstrapper.js [hunk 10]
```diff
@@ -86,8 +105,13 @@ export default class Bootstrapper {
     async createRunnableConfiguration () {
         var reporterPlugin = this._getReporterPlugin();
 
-        var tests      = await this._getTests();
-        var browserSet = await this._getBrowserConnections();
+        // NOTE: If user forgot to specify a browser, but has specified a path to tests, the specified path will be
+        // considered as the browser argument, and the tests path argument will have predefined default value.
+        // It's very ambiguous for the user, which might be confused by compilation errors from an unexpected test.
+        // So we need to retrieve browser aliases and paths before tests compilation.
```
Comment (VasilyStrelyaev) [inline src/runner/bootstrapper.js]:
```
        // NOTE: If a user forgot to specify a browser, but has specified a path to tests, the specified path will be
        // considered as the browser argument, and the tests path argument will have the predefined default value.
        // It's very ambiguous for the user, who might be confused by compilation errors from an unexpected test.
        // So, we need to retrieve the browser aliases and paths before tests compilation.
```

**Comment by VasilyStrelyaev** [comment]
\r-

File: src/cli/argument-parser.js [hunk 11]
```diff
@@ -85,6 +85,49 @@ export default class CLIArgumentParser {
         return '\n' + wordWrap(DESCRIPTION, 2, getViewPortWidth(process.stdout));
     }
 
+    static _parseBrowserArg (browserArg) {
+        var currentBrowser = '';
+        var browsers       = [];
+
+        // NOTE: quotedChar contains ' or " - starting character of quotation when we are inside an enquoted part,
+        // or an empty string otherwise.
+        var quoteChar = '';
```
Comment (inikulin) [inline src/cli/argument-parser.js]: Let's use `null` instead of `''`. Then you will not need this comment.

File: src/cli/argument-parser.js [hunk 12]
```diff
@@ -85,6 +85,49 @@ export default class CLIArgumentParser {
...
+                break;
+
+            case '"':
+            case '\'':
+                if (quoteChar === currentChar)
+                    quoteChar = '';
+                else if (!quoteChar)
+                    quoteChar = currentChar;
+
+                currentBrowser += currentChar;
```
Comment (inikulin) [inline src/cli/argument-parser.js]: Is ther any need to include quote char into the parsed browser name?

Comment (AndreyBelym) [inline src/cli/argument-parser.js]: My thought is that there is a a point of including quote char only if it is escaped by another type of quotes, e.g. `path:"/John's folder/firefox.exe"`. `'` is valid for file names even on Windows, `"` is valid for UNIX'es.

**Comment by AndreyBelym** [comment]
FPR

File: src/cli/argument-parser.js [hunk 13]
```diff
@@ -85,6 +85,47 @@ export default class CLIArgumentParser {
         return '\n' + wordWrap(DESCRIPTION, 2, getViewPortWidth(process.stdout));
     }
 
+    static _parseBrowserArg (browserArg) {
+        var currentBrowser = '';
+        var browsers       = [];
+
```
Comment (inikulin) [inline src/cli/argument-parser.js]: Unnecessary new line

**Comment by inikulin** [comment]
Overall lgtm, just a minor style issue

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
lgtm

**Pull request merged**